lib/elixir/lib/kernel/typespec.ex

```diff
@@ -598,28 +616,26 @@ defmodule Kernel.Typespec do
 
         # This may be generating an invalid typespec but we need to generate it
         # to avoid breaking existing code that was valid but only broke dialyzer
-        {left, state} = typespec(var, [elem(var, 0) | vars], caller, state)
         {right, state} = typespec(expr, vars, caller, state)
-        {{:ann_type, line(meta), [left, right]}, state}
+        {{:ann_type, line(meta), [{:var, line(var_meta), var_name}, right]}, state}
 
       {right, state} ->
-        {left, state} = typespec(var, [var_name | vars], caller, state)
-        {{:ann_type, line(meta), [left, right]}, state}
+        {{:ann_type, line(meta), [{:var, line(var_meta), var_name}, right]}, state}
     end
   end
 
-  defp typespec({:::, meta, [var, expr]} = ann_type, vars, caller, state) do
+  defp typespec({:::, meta, [left, right]} = expr, vars, caller, state) do
     message =
       "invalid type annotation. When using the | operator to represent the union of types, " <>
-        "make sure to wrap type annotations in parentheses: #{Macro.to_string(ann_type)}"
+        "make sure to wrap type annotations in parentheses: #{Macro.to_string(expr)}"
 
     # TODO: make this an error in elixir 2.0 and remove code below
     :elixir_errors.warn(caller.line, caller.file, message)
 
     # This may be generating an invalid typespec but we need to generate it
     # to avoid breaking existing code that was valid but only broke dialyzer
-    {left, state} = typespec(var, [elem(var, 0) | vars], caller, state)
```

Comment:
This clause is meant for invalid type annotations due to the operator precedence between the type operator `::` and the pipe operator `|`. For example, `@type my_type :: pid | port | registered_name :: atom`.

Because of that, adding `elem(var, 0)` to the list of `vars` makes no sense because it's going to be the pipe operator `:|`.

On top of that, this kind of invalid type annotations will always raise a `CompileError` because the left part is never going to be a valid typespec. For instance, in the example above, the left part `pid | port | registered_name` would raise `type registered_name/0 undefined`.

So, should we get back to raising a `CompileError` here or should we try to find a workaround?

cc @ericmj

Reply:
If it would cause a compile error before then it's okay to still raise a compile error. But we can't introduce new compile errors like the other PR did.

Reply:
I think I can manipulate the AST and convert `pid | port | registered_name :: atom` into `pid | port | (registered_name :: atom)`, and generate a valid typespec. 

I think we can even remove the warning, and include a new one in case of ambiguous type annotations. So, `pid | port | registered_name :: atom` won't emit a warning, but `pid | port | registered_name :: atom | String.t()` will.

Reply:
@josevalim the main issue is that the `CompileError` is of the type "undefined type", because it parses the name of the type annotation as a user type.

To give an example, in estable Elixir versions this type definition: 

```
@type dest :: pid | port | registered_name :: atom | {atom, node}
``` 

is passed as:

```elixir
 [
   type: {:dest,
    {:ann_type, 2,
     [
       {:type, 2, :union,
        [
          {:type, 2, :pid, []},
          {:type, 2, :port, []},
          {:user_type, 2, :registered_name, []}
        ]},
       {:type, 2, :union,
        [
          {:type, 2, :atom, []},
          {:type, 0, :tuple, [{:type, 2, :atom, []}, {:type, 2, :node, []}]}
        ]}
     ]}, []}
 ]
```

but surprisingly Erlang does not raise a `CompileError: type registered_name() undefined` on `{:user_type, 2, :registered_name, []}` (nor a syntax error). 

In https://github.com/elixir-lang/elixir/pull/8163, we moved that check to Elixir land, and since then we are raising a `CompileError: type registered_name/0 undefined` for that case. 

I thought about simulating the old behavior by including temporary the name of the type annotation to the list of `defined_type_pairs`, so it does not raise in here:

https://github.com/elixir-lang/elixir/blob/e1e096ab2ed1b705926b3e05a397d3d1a40d2977/lib/elixir/lib/kernel/typespec.ex#L766-L768

But, that wouldn't work for cases like `@type my_type :: pid | port | {node, name} | name :: atom`.

That's why the only option I see is to manipulate the AST to generate the proper typespec. In fact, I think this option is better than the original `CompileError: invalid type annotation` because: 

* it's compatible with older versions of Elixir and it fixes the bad generated typespecs. Users might get their Dialyzer errors fixed just by upgrading to 1.8 🎉 
* it turns `pid | port | registered_name :: atom` into a valid expression, which I think it's right since there is not ambiguity nor invalid syntax.
* it would handle these three typespecs as equivalent:

  * `pid | port | registered_name :: atom | {atom, node}`
  * `pid | port | (registered_name :: atom) | {atom, node}`
  * `pid | port | (registered_name :: atom | {atom, node})`

  And only emit warning due to the ambiguity of the type annotation.

Thoughts?

Reply:
I think trying to fix syntax after the fact is a very dangerous slippery slope because we may end-up adding more ambiguity that we will have so solve later on. In particular, those last two specs are definitely not equivalent:

```elixir
pid | port | (registered_name :: atom) | {atom, node}
pid | port | (registered_name :: atom | {atom, node})
```

As I would expect the second one to be equivalent to `pid | port | (registered_name :: (atom | {atom, node}))`.

Reply:
Sorry, I should have clarified that I meant "equivalent" in the sense of the types Dyalizer will see. That's why it will emit a warning due to the ambiguity of the type annotation.

Reply:
In addition, I should also have clarified that I only see these two alternatives:

* change the code back to raise a `CompileError: invalid type annotation`
* manipulate the AST

because I think we cannot keep the `warning: invalid type annotation` and circumvent the `CompileError: type .. undefined` (for the reasons explained in the comment above).

Reply:
My suggestion then:

1. Let's add the warning about this being a wrong contruction to Elixir v1.7 branch. We will do a new release and people will start fixing it today.

2. Let's postpone merging this until after v1.8 is out. This means it will be part of v1.9. We will be without the "variable used once" warning for a while but we will survive.

@fertapric what do you think?

Reply:
@josevalim this conversation is orthogonal to the "variable used one" (singleton type var) error. Sorry, I mentioned it here because I discovered it while working on this PR. They are apples and oranges!

In order for this to work, I think we should remove the check for undefined types introduced in https://github.com/elixir-lang/elixir/pull/8163, and delegate that check to `erl_lint` until v1.9.

I would like to clarify the manipulation of the AST I'm suggesting:

Having this typespec `pid | port | registered_name :: atom`, with the AST

```elixir
{:::, [],
 [
   {:|, [],
    [
      {:pid, [], Elixir},
      {:|, [],
       [{:port, [], Elixir}, {:registered_name, [], Elixir}]}
    ]},
   {:atom, [], Elixir}
 ]}
```

it would detect the right-most pipe argument and replace it with the type annotation expression:

```elixir
{:|, [],
 [
   {:pid, [], Elixir},
   {:|, [],
    [{:port, [], Elixir}, 
     {:::, [], [{:registered_name, [], Elixir}, {:atom, [], Elixir}]}]}
 ]}
```

The code should be something like:

```elixir
defp rewrite_ann_type({:::, meta, [left, right]}) do
  do_rewrite_ann_type(meta, right, left)
end

defp do_rewrite_ann_type(ann_type_meta, ann_type_expr, {:|, meta, [left, {:|, _, _} = right]}) do
  right = do_rewrite_ann_type(ann_type_meta, ann_type_expr, right)
  {:|, meta, [left, right]}
end

defp do_rewrite_ann_type(ann_type_meta, ann_type_expr, {:|, meta, [left, right]}) do
  right = {:::, ann_type_meta, [right, ann_type_expr]}
  {:|, meta, [left, right]}
end
```

Reply:
We shouldn't do AST manipulation because we will fix the precedence in just some cases instead of giving feedback to the user that the precedence is wrong. I am a 100% it will backfire in one way or the other.

If this commit is not related to the warning and the warning works fine today, what am I missing?

> In order for this to work, I think we should remove the check for undefined types introduced in #8163

What does "this" refer to in this case? This PR? The feature? Can we have a test case for what is currently failing that is warranting this discussion?

Reply:
Ah, I see now. The undefined_type check made the warning obsolete. I think what erlang is doing is that it is considering everything on the left-side of `::` to be "defined". Maybe we can do the same change here temporarily? We could do this by storing a flag on the state that says that we are on the left side of a `::`.

Reply:
In any case, I understand that this is a topic of another PR, let's focus on this one and then tackle the other. Can you please open up an issue?

Reply:
Yeah. Here is the issue https://github.com/elixir-lang/elixir/issues/8235

Sorry for all the noise generated in this PR. I thought the discussion was going to be much sorter! 😓